testapp/testapp/AppDelegate.swift

```diff
@@ -0,0 +1,156 @@
+//
+//  AppDelegate.swift
+//  testapp
+//
+//  Created by Christopher Batin on 16/01/2019.
+//  Copyright © 2019 Christopher Batin. All rights reserved.
+//
+
+import UIKit
+import Ably
+import UserNotifications
+
+// 1
+let authURL = "YOUR_NGROK_HTTPS/auth"
+@UIApplicationMain
+class AppDelegate: UIResponder, UIApplicationDelegate, ARTPushRegistererDelegate {
+    var window: UIWindow?
+    var realtime: ARTRealtime!
+    var channel: ARTRealtimeChannel!
+    var subscribed = false
+
+    //2
+    func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
+        DispatchQueue.main.async() {
+            print("** didRegisterForRemoteNotificationsWithDeviceToken")
+            ARTPush.didRegisterForRemoteNotifications(withDeviceToken: deviceToken, realtime: self.getAblyRealtime())
+        }
+    }
+
+    func application(_ application: UIApplication, didFailToRegisterForRemoteNotificationsWithError error: Error) {
+        DispatchQueue.main.async() {
+            print("** didFailToRegisterForRemoteNotificationsWithError")
+            ARTPush.didFailToRegisterForRemoteNotificationsWithError(error, realtime: self.getAblyRealtime())
+        }
+    }
+
+    //3
+    private func getAblyRealtime() -> ARTRealtime {
+        let options = ARTClientOptions()
+        options.authCallback = { params, callback in
+            self.getTokenRequest() { json, error in
+                do {
+                    callback(try ARTTokenRequest.fromJson(json!), nil)
+                } catch let error as NSError {
+                    callback(nil, error)
+                }
+            }
+        }
+        realtime = ARTRealtime(options: options)
+        realtime.connection.on { state in
+            if let state = state {
+                switch state.current {
+                case .connected:
+                    print("connected")
+                case .failed:
+                    print("failed")
+                default:
+                    break
+                }
+            }
+        }
+        return realtime
+    }
+
+
+    // 4
+    func didActivateAblyPush(_ error: ARTErrorInfo?) {
+        if let error = error {
+            // Handle error
+            print("** push activation failed, err=\(String(describing: error))")
+            return
+        }
+        print("** push activated")
+
+        self.channel = self.realtime.channels.get("push")
+        self.channel.attach() { (err) in
+            print("** channel attached, err=\(String(describing: err))")
+
+            self.channel.push.subscribeDevice { (err) in
+                DispatchQueue.main.async {
+                    print("** device ID \(self.realtime.device.id)")
+                }
+                print("** channel.push.subscribeDevice: err=\(String(describing: err))")
+                self.subscribed = true
```

Comment:
@cjbatin Shouldn't this also be synced with `DispatchQueue.main`?

Reply:
It makes no difference. The only reason print is in the Dispatch is because I needed it displaying in the console

Reply:
I still don't understand this. Sorry. It should work without `Dispatch`.